dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorArm8.java

```diff
@@ -223,4 +239,291 @@ public List<Event> visitDat3mCAS(Dat3mCAS e) {
+	@Override
+	public List<Event> visitLKMMLoad(LKMMLoad e) {
+		Register resultRegister = e.getResultRegister();
+		IExpr address = e.getAddress();
+		String mo = e.getMo();
+		
+        Load load = newLoad(resultRegister, address, ARMv8.extractLoadMoFromLKMo(mo));
+        
+        return eventSequence(
+                load
+        );
+	}
+
+	// Following
+	//		https://elixir.bootlin.com/linux/v5.18/source/arch/arm64/include/asm/barrier.h#L116
+	@Override
+	public List<Event> visitLKMMStore(LKMMStore e) {
+		ExprInterface value = e.getMemValue();
+		IExpr address = e.getAddress();
+		String mo = e.getMo();
+
+        Store store = newStore(address, value, mo.equals(Tag.Linux.MO_RELEASE) ? Tag.ARMv8.MO_REL : null);
+        
+        return eventSequence(
+                store
+        );
+	}
+
+	// Following
+	//		https://elixir.bootlin.com/linux/v5.18/source/arch/powerpc/include/asm/barrier.h
+	@Override
+	public List<Event> visitLKMMFence(LKMMFence e) {
+		Fence optionalMemoryBarrier;
+		switch(e.getName()) {
+			// mb()
+			case Tag.Linux.MO_MB:
+				optionalMemoryBarrier = AArch64.DSB.newISHBarrier();
+				break;
+			// rmb()
+			case Tag.Linux.MO_RMB:
+				optionalMemoryBarrier = AArch64.DSB.newISHLDBarrier();
+				break;
+			// wmb()
+			case Tag.Linux.MO_WMB:
+				optionalMemoryBarrier = AArch64.DSB.newISHSTBarrier();
+				break;
+			// __smp_mb()
+			// 		https://elixir.bootlin.com/linux/v5.18/source/include/asm-generic/barrier.h
+			case Tag.Linux.BEFORE_ATOMIC:
+			case Tag.Linux.AFTER_ATOMIC:
+				optionalMemoryBarrier = AArch64.DMB.newISHBarrier();
+				break;
+			default:
+				optionalMemoryBarrier = null;
+				break;
+		}
+
+		return eventSequence(
+                optionalMemoryBarrier
+        );
+	}
+	
+	// =============================================================================================
+	// 										GENERAL COMMENTS
+	// =============================================================================================
+	// We currently only support LL/SC (exclusive load/store) compilation. 
+	// However the kernel also supports using hardware atomic operations
+	// 		https://elixir.bootlin.com/linux/v5.18/source/arch/arm64/include/asm/lse.h
+	// =============================================================================================
+
+	// Following
+	// 		https://elixir.bootlin.com/linux/v5.18/source/arch/arm64/include/asm/atomic_ll_sc.h#L259
+	// Instead of using a comparison, it uses a XOR
+	@Override
+	public List<Event> visitRMWCmpXchg(RMWCmpXchg e) {
+		Register resultRegister = e.getResultRegister();
+		IExpr address = e.getAddress();
+		ExprInterface value = e.getMemValue();
+		String mo = e.getMo();
+
+		Register dummy = e.getThread().newRegister(e.getResultRegister().getPrecision());
+        Label casEnd = newLabel("CAS_end");
+        CondJump branchOnCasCmpResult = newJump(new Atom(new IExprBin(dummy, IOpBin.XOR, (IExpr) e.getCmp()), NEQ, IValue.ZERO), casEnd);
```

Comment:
Are you sure that we want to mimic the real compilation here? I think the `XOR` can harm performance substantially without affecting the semantics at all.

Reply:
I think when there is inline assembly involved, we should be as close as possible to the real implementation. If performance is really an issue, I agree we should use a semantic equivalent, but cheaper, model (at the end of the day, we have several "modelling assumptions"). However I have been testing this quite a lot with real code from the kernel and performance is not an issue at all.

Reply:
Just to make sure: Did you test this with integer encoding or bit vector encoding?
When using BV-encoding, there is no issue at all cause a XOR is cheap. If not, then this causes a conversion from IA to BV, which can always be dangerous.
In some SVCOMP benchmarks, just having a boolean negation implement via `XOR 1 `already caused serious issues in the past to the point. So I would be surprised if something like this doesn't happen here as well.

Reply:
Fair question ... when I mentioned that I tested this with real code, I was always using the BV-encoding. Since we don't have any real code from the kernel not suing BVs (i.e. where only the compilation would introduce bit-precise reasoning), I did the following: I changed all other encodings of `IOpBin` operators to just use random integer operations (I know this is completely unsound, I just wanted to measure the performance impact of adding a single bit-precise operation) and test the compilation on qspinlock. You are right, the overhead is considerable (~10X) thus I will change the compilation to use `==` instead or `xor` and just keep a comment about the rationale of the difference between the real compilation and the encoding.